src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,82 @@
 package mate.academy.lib;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.util.HashMap;
+import java.util.Map;
+import mate.academy.service.FileReaderService;
+import mate.academy.service.ProductParser;
+import mate.academy.service.ProductService;
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
     private static final Injector injector = new Injector();
+    private final Map<Class<?>, Object> instances = new HashMap<>();
 
     public static Injector getInjector() {
         return injector;
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        Object clazzImplementationInstance = null;
+        Class<?> clazz = findImplementation(interfaceClazz);
+        Field[] declaredFields = clazz.getDeclaredFields();
+        for (Field field: declaredFields) {
+            if (field.isAnnotationPresent(Inject.class)) {
+                Object fieldInstance = getInstance(field.getType());
+                try {
+                    clazzImplementationInstance = createNewInstance(clazz);
+                } catch (Exception e) {
+                    throw new RuntimeException("Injection failed, missing @Component annotaion "
+                            + "on the class " + clazz.getName(), e);
+                }
+                field.setAccessible(true);
+                try {
+                    field.set(clazzImplementationInstance, fieldInstance);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException("Can't initialize field value. "
+                            + "Class: " + clazz.getName() + ". Field: " + field.getName());
+                }
+            }
+        }
+        if (clazzImplementationInstance == null) {
+            try {
+                clazzImplementationInstance = createNewInstance(clazz);
+            } catch (Exception e) {
+                throw new RuntimeException("Injection failed, missing @Component annotaion "
+                        + "on the class " + clazz.getName(), e);
+            }
+        }
+        return clazzImplementationInstance;
+    }
+
+    private Object createNewInstance(Class<?> clazz) throws Exception {
+        if (instances.containsKey(clazz)) {
+            return instances.get(clazz);
+        }
+        if (clazz.isAnnotationPresent(Component.class)) {
```

Comment:
add this check in getInstance() method before creation instance, and throw RuntimeException with message if it is false. And after that you can delete redundant try-catch block inside getInstance method